Riot/ViewController/RoomViewController.m

```diff
@@ -4600,29 +4600,49 @@ - (void)contactsTableViewController:(ContactsTableViewController *)contactsTable
 
 #pragma mark - Re-request encryption keys
 
-- (void)showRerequestConfirmationAlert:(MXEvent*)event
+- (void)reRequestKeysAndShowExplanationAlert:(MXEvent*)event
 {
-    currentAlert = [UIAlertController alertControllerWithTitle:NSLocalizedStringFromTable(@"rerequest_keys_alert_title", @"Vector", nil)
+    MXWeakify(self);
+    __block UIAlertController *alert;
+
+    // Make the re-request
+    [self.mainSession.crypto reRequestRoomKeyForEvent:event];
+
+    // Observe kMXEventDidDecryptNotification to remove automatically the dialog
+    // if the user has shared the keys from another device
+    id didDecryptObserver;
+    didDecryptObserver = [[NSNotificationCenter defaultCenter] addObserverForName:kMXEventDidDecryptNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *notif) {
```

Comment:
You should store `didDecryptObserver` as instance variable and remove it from NSNotificationCenter in `dealloc`. Otherwise this observer block could never be removed if controller is deallocated and "ok" alert not tapped.